Content.Shared/Chat/V2/Moderation/SimpleCensor.cs

```diff
@@ -0,0 +1,272 @@
+
+        var censored = input.ToCharArray();
+
+        input = Sanitize(input, out var originalIndexes, true);
+        CheckProfanity(ref input, ref originalIndexes, ref censored, _falseNegatives, replaceWith);
+        RemoveFalsePositives(ref input, ref originalIndexes);
+        CheckProfanity(ref input, ref originalIndexes, ref censored, _censoredWords, replaceWith);
+
+        return new string(censored);
+    }
+
+    private void CheckProfanity(ref string input, ref List<int> originalIndexes, ref char[] censored, List<string> words, char replaceWith = '*')
+    {
+        foreach (var word in words)
+        {
+            var wordLength = word.Length;
+            var currentIndex = 0;
+
+            for (var foundIndex = input[currentIndex..].IndexOf(word, StringComparison.OrdinalIgnoreCase); foundIndex > -1; foundIndex = input[currentIndex..].IndexOf(word, StringComparison.OrdinalIgnoreCase))
+            {
+                for (var j = 0; j < wordLength; j++)
+                {
+                    var runeIdx = IndexToRune(input, currentIndex + foundIndex) + j;
+                    if (runeIdx < originalIndexes.Count)
+                    {
+                        censored[originalIndexes[runeIdx]] = replaceWith;
+                    }
+                }
+
+                currentIndex += foundIndex + wordLength;
+            }
+        }
+    }
+
+    private void RemoveFalsePositives(ref string input, ref List<int> originalIndexes)
+    {
+        foreach (var word in _falsePositives)
+        {
+            var wordLength = word.Length;
+            var currentIndex = 0;
+
+            for (var foundIndex = input[currentIndex..].IndexOf(word, StringComparison.OrdinalIgnoreCase); foundIndex > -1; foundIndex = input[currentIndex..].IndexOf(word, StringComparison.OrdinalIgnoreCase))
+            {
+                // shufflin'
+                var foundRuneIndex = IndexToRune(input, foundIndex);
+                var newOriginalIndexes = new List<int>(originalIndexes[..foundRuneIndex]);
+                newOriginalIndexes.AddRange(originalIndexes[(foundRuneIndex + wordLength)..]);
+                originalIndexes = newOriginalIndexes;
+
+                currentIndex += foundIndex + wordLength;
+            }
+
+            input = input.Replace(word, "");
+        }
+    }
+
+    private string Sanitize(string input, out List<int> originalIndexes, bool rememberOriginalIndexes)
+    {
+        input = input.ToLower();
+        if (_sanitizeL33TSpeak && _sanitizeSpecialCharacters)
+        {
+            // catch one specific trick
+            input = input.Replace("()", "o");
+        }
+
+        foreach (var rune in input)
+        {
+            if (!_characterReplacements.TryGetValue(rune, out var repl))
+                continue;
+
+            if (_sanitizeSpecialCharacters && repl == ' ')
+            {
+                input = input.Replace(rune, ' ');
```

Comment:
Doing replace operations like this has to fully copy the string which is very bad for performance, i.e. this code is O(n^2).